src/prompt/functions/ai-chain.ts

```diff
@@ -0,0 +1,174 @@
+    >;
+    prompt?: Prompt.Template<Params>;
+    functions?: Prompt.AIFunction[];
+    maxCalls?: number;
+    maxRetries?: number;
+    toolCallConcurrency?: number;
+  },
+  context?: Model.Ctx
+): Prompt.Chain<Params, Result> {
+  return async (promptArgs: Params): Promise<Result> => {
+    if (!params?.messages && !prompt) {
+      throw new Error('Either params.messages or prompt must be provided');
+    }
+
+    if (params?.messages && prompt) {
+      throw new Error('Only one of params.messages or prompt may be provided');
+    }
+
+    const messages =
+      params?.messages ?? (await Promise.resolve(prompt!(promptArgs)));
+    const tools = functions?.map((func) => ({
+      type: 'function' as const,
+      function: func.spec,
+    }));
+
+    let numCalls = 0;
+    let numErrors = 0;
+
+    while (numCalls < maxCalls) {
+      ++numCalls;
+      const response = await chatModel.run(
+        {
+          ...params,
+          messages,
+          tools,
+        },
+        context
+      );
+
+      const { message } = response;
+      try {
+        if (Msg.isToolCall(message)) {
+          messages.push(message);
+
+          if (!functions) {
+            throw new AbortError('No functions provided to handle tool call');
+          }
+
+          await pMap(
+            message.tool_calls,
+            async (toolCall) => {
+              const func = functions?.find(
+                (func) => func.spec.name === toolCall.function.name
+              );
+
+              if (!func) {
+                throw new Error(
+                  `No function found with name ${toolCall.function.name}`
+                );
+              }
+
+              // TODO: ideally we'd differentiate between tool argument validation
+              // errors versus errors in the function implementation
+              const result = await func(toolCall.function.arguments);
+
+              const toolResult = Msg.toolResult(
+                JSON.stringify(result, null, 2),
+                toolCall.id
+              );
+
+              messages.push(toolResult);
+            },
+            {
+              concurrency: toolCallConcurrency,
+            }
+          );
+        } else if (Msg.isFuncCall(message)) {
+          throw new AbortError(
+            'Function calls are not supported; expected tool call'
+          );
+        } else if (Msg.isAssistant(message)) {
+          if (schema) {
+            // TODO: we're passing a schema, but we're not actually telling the
+            // chat model anything about the expected output format
+            if (schema instanceof z.ZodObject) {
+              return extractZodObject({ json: message.content, schema });
+            } else {
+              // TODO: support arrays, boolean, number, etc.
+              // validate string output
+              return schema.parse(message.content);
+            }
+          } else {
+            return message.content as Result;
+          }
+        }
```

Comment:
Given that we aren't actually telling the model about this and most of these chains will terminate in many different ways, I don't think it makes sense for this to take on the complexity of parsing the response. It will be much cleaner this this always returns a string and parsing can be done on the chains output. Am I missing something here?

Reply:
hmmm you wanted this to replace `runWithValidation`, and the only thing that really did was the output validation / parsing w/ retries, so that's why I have this in here.

Reply:
I could go either way with this for `createAIChain`, but we need some built-in ability to parse structured output with retries.

Reply:
One option would be to tell the model about the expected output if a `schema` is passed via an additional `system` message.

Reply:
i didnt realize this was to replace `runWithValidation`. i dont think the validation component is the same as this, which i see as being responsible for making the calls and validating the function arguments.

what do you think of having a separate validator that is meant to go after the model generates a response?

Reply:
an added benefit of validating in the loop like this is that its much easier to rerun with the messages for context than if you fully break out of the loop.